ksqldb-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonSerdeFactory.java

```diff
@@ -127,6 +128,7 @@ private static Converter getSchemaConverter(
         ksqlConfig.getString(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY)
     );
     config.put(JsonConverterConfig.DECIMAL_FORMAT_CONFIG, DecimalFormat.NUMERIC.name());
+    config.put(JsonSchemaDataConfig.OBJECT_ADDITIONAL_PROPERTIES_CONFIG, false);
```

Comment:
I haven't dug into the code, but what happens if we're using a schema that already has it explicitly set to `true`? will our schema override that? (e.g. does this make it impossible to `INSERT INTO` a source that uses JSON_SR with the value for true?) 

Also what does this mean for backwards compatibility with existing streams/tables that were created with the config being false. Can you test `CREATE STREAM AS SELECT ...` from a `JSON_SR` stream before upgrading, upgrade, and then see if we can insert and read from the stream?

Reply:
Thanks for taking a look, @agavra!

Given `additionalProperties=true`, the schema can never be evolved, even when the updated schema sets `additionalProperties=false`.

> Also what does this mean for backwards compatibility with existing streams/tables that were created with the config being false. Can you test CREATE STREAM AS SELECT ... from a JSON_SR stream before upgrading, upgrade, and then see if we can insert and read from the stream?

With a stream created before this change, upgrade then insert will never work. Unfortunately ksqlDB cannot feasibly fix forward existing queries with schemas created before this change because of the way JsonSchema works. I found @rayokota's post useful for understanding the behavior: https://yokota.blog/2021/03/29/understanding-json-schema-compatibility/.

Reading from the upgraded stream would work because no schema is written on CREATE OR UPDATE, and, if I understand correctly, the existing reader schema is loaded based on the version in the record header.

Reply:
I don't think my intention came across clearly - I meant does this affect existing queries when upgrading ksqlDB versions, without issuing a `CREATE OR REPLACE`?

Reply:
Ah sorry for misinterpreting. My understanding is the change would not affect existing queries. We only register schemas on CREATE and INSERT, and this is the only point where the new property comes into play.

Reply:
We register schemas every time we serialize something (if the schema differs) - so it might still affect existing queries if we are serializing into a topic. What I'm worried about is that the converter is not configured in the physical plan, so we'd suddenly start using this serializer for old queries as well.

It might be better to hop on a quick zoom to see if one of us is misunderstanding the other 😆

Reply:
Thanks for checking me on this.

I had missed the implications of upgrading ksqlDB itself upon existing JSON_SR-based queries. The change would indeed cause the converter to attempt to serialize an incompatible schema. Unfortunately, while the effect is essentially to add `additionalProperties: false` to the schema, this is not a backwards-compatible operation against an existing schema that does not set this property.

One recourse is to configure the converter based on a new property in KsqlConfig, documenting opting in to the `additionalProperties=false` as the best practice.